File: leetcode/python/Remove Duplicates from Sorted List II/Remove_Duplicates_from_Sorted_List_II.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    # @param head, a ListNode
    # @return a ListNode
    def deleteDuplicates(self, head):
        d = {}
        h = head
        while h:
            if h.val not in d:
                d[h.val] = 1
            else:
                d[h.val] += 1
            h = h.next
        res = ListNode(0)
        end = res
        while head:
            if d[head.val] == 1:
                end.next = head
                end = end.next
            head = head.next
        end.next = None
        return res.next
            

File: leetcode/cpp/RemoveDuplicatesfromSortedListII.cpp
//
// Created by zz on 2017/10/30.
// Time:  O(n)
// Space: O(1)

struct ListNode {
    int val;
    ListNode *next;

    ListNode(int x) : val(x), next(nullptr) {}
};

class Solution {
public:
    ListNode *deleteDuplicates(ListNode *head) {
        ListNode dummy{0};
        ListNode *curr = head;
        ListNode *prev = &dummy;
        while (curr) {
            if (curr->next && curr->val == curr->next->val) {
                int val = curr->val;
                while (curr && curr->val == val) {
                    curr = curr->next;
                }
                prev->next = curr;
            } else {
                prev->next = curr;
                prev = curr;
                curr = curr->next;
            }
        }
        return dummy.next;
    }
};